README.md
# Golden Parameter Search (GPS)

Golden Parameter Search (GPS) [Pushak & Hoos, 2020] is an automated 
algorithm congifuration procedure. That is, it seeks to optimize the 
performance of a target algorithm on a set of instances by automatically 
finding high-quality values for the parameters of the target algorithm. GPS 
can optimze the performance of an algorithm in terms of either the running
time required to solve problem instances or the solution quality found by
the target algorithm for the problem instances.
 
GPS is the first automated algorithm configuration procedure to exploit
recent insights into the structural properties of algorithm configuration
landscapes [Pushak & Hoos, 2018]. In particular, GPS assumes that each
parameter of the target algoritm effects a uni-modal response in the 
performance of the algorithm, when modified individually. Furthermore,
GPS assumes that most parameters of the target algorithm do not strongly
interact, thereby allowing GPS to optimize each parameter semi-independently
in parallel. These two assumptions allow GPS to quickly and efficiently explore
the parameter configuration space. However, if you have reason to believe that
your particular target algorithm contains parameters that strongly violates
either of these two assumptions, then GPS may not be the appropriate algorithm
configuration procedure for you. 

If you use GPS in your work, please consider citing our paper.

 - Yasha Pushak and Holger H. Hoos.  
**Golden Parameter Search: Exploiting Structure to Quickly Configure Parameters
In Parallel.**  
*In Proceedings of the Twenty-Second Interntional Genetic and Evolutionary 
Computation Conference (GECCO 2020)*. pp 245-253 (2020).  
**Won the 2020 GECCO ECOM Track best paper award.**
 - Yasha Pushak and Holger H. Hoos.  
**Algorithm Configuration Landscapes: More Benign than Expected?**  
*In Proceedings of the Fifteenth Internationl Conference on Parallel Problem 
Solving from Nature (PPSN 2018)*. pp 271-283 (2018).  
**Won the 2018 PPSN best paper award.**

# Table of Contents

   * [Golden Parameter Search (GPS)](#golden-parameter-search-gps)
   * [Table of Contents](#table-of-contents)
   * [Installing GPS](#installing-gps)
      * [Setting up a Redis Database](#setting-up-a-redis-database)
   * [Quick Start Guide](#quick-start-guide)
      * [Required Input](#required-input)
      * [Example command line call for GPS](#example-command-line-call-for-gps)
      * [Experiment Directory](#experiment-directory)
      * [Using a Scenario file](#using-a-scenario-file)
      * [Temporary File Directory - <strong>Important</strong>](#temporary-file-directory---important)
      * [Solution Quality Optimization](#solution-quality-optimization)
   * [Target Algorithm Wrapper](#target-algorithm-wrapper)
      * [Command Line Target Algorithm Wrapper Format](#command-line-target-algorithm-wrapper-format)
         * [Target Algorithm Wrapper Input](#target-algorithm-wrapper-input)
         * [Target Algorithm Wrapper Output](#target-algorithm-wrapper-output)
      * [Python Target Algorithm Wrapper Interface](#python-target-algorithm-wrapper-interface)
   * [Instance File Format](#instance-file-format)
   * [Parameter Configuration Space File Format](#parameter-configuration-space-file-format)
      * [Conditional Parameters](#conditional-parameters)
      * [Example Configuration Space](#example-configuration-space)
      * [Forbidden Statements](#forbidden-statements)
      * [Old Parameter Configuration Space Syntax](#old-parameter-configuration-space-syntax)
   * [GPS Arguments](#gps-arguments)
      * [Setup Arguments](#setup-arguments)
      * [Redis Arguments](#redis-arguments)
      * [Scenario Arguments](#scenario-arguments)
      * [GPS Parameters](#gps-parameters)
      * [Post-Process Parameters](#post-process-parameters)
   * [Contact](#contact)

# Installing GPS

 1. Create a python2.7 virtual environment
 2. Download the latest version of the parameter configuration space parser
from https://github.com/YashaPushak/PCS 
 3. While in the main PCS directory, install PCS with 
`pip install .`
or
`python setup.py install --user`.
 4. Download the latest version of GPS from https://github.com/YashaPushak/GPS
 5. While in the main GPS directory, install GPS's other required python 
packages
`pip install -r requirements.txt`.
 6. While in the main GPS directory, install GPS with 
`pip install .`
or 
`python setup.py install --user`
 7. Setup and install a [redis database](#setting-up-a-redis-database). 

## Setting up a Redis Database

These instructions help you to setup a simple local database.
You can find more information on redis on their official website: [redis.io](https://redis.io).
The configuration file provided is only a light modification of [the one provided with redis 5.0](https://raw.githubusercontent.com/antirez/redis/5.0/redis.conf).

You first need to install redis. It is available on most systems, as well as on Anaconda using:

    conda install redis 

To start local a server, use the following command:

    redis-server ./redis/redis.conf

This will start a server that listens to port 9503 on your local machine. It has 16 databases with `dbid` from 0 to 15.

In the file `redis_configuration.txt` change `redis-host` to `localhost`

You should now be ready to run GPS.

# Quick Start Guide

GPS is implemented in python2.7, but is designed to be used from the command
line. To use GPS you will need to launch a master process and one or more
worker processes. The master process will loop through each parameter, check
for newly completely target algorithm runs and queue new target algorithm
runs to be performed. The worker processes will repeatedly check for new
target algorithms, perform them and then save the reuslts. Communication 
between the master and worker processes is done through a redis database.

## Required Input

To use GPS, you must provide it with several minimum requirements. We will
use the artificial algorithm example scenario provided with GPS as a running
example. This artificial algorithm hallucinates running times for simulated
parameter responses, all of which is designed to approximate the behaviour
of real target algorithms. The scenario was designed to be an easy benchmark 
for GPS that can be run in less than 5 minutes with 2 processors (it typically
takes between 60-90 seconds on our machines).

GPS requires several pieces of information about your scenario to run:
 
**Target Algorithm:** A target algorithm to optimize. This corresponds to the
GPS's `algo` argument. For example: 
`--algo 'python2 examples/artificial-algorithm/algorithm.py'`. The target
algorithm can either be callable via the command line, or it can implement
a simple python interface.
See [Target Algorithm Wrapper](#target-algorithm-wrapper) for more details.

**Instance File:** A file that specifies the instances on which your your target
algorithm should be evaluated. Each line should contain a single instance name.
This corresponds to the `instance-file` argument. For example:
`--instance-file examples/artificial-algorithm/instances.txt`
See [Instance File Format](#instance-file-format) for more details.

**Parameter Configuration Space File:** A parameter configuration space (.pcs) 
file, which defines the parameters of your target algorithm that GPS should 
optimize, including their names, (suggested) domains and default values. This 
corresponds to the `pcs-file` argument. For example:
`--pcs-file examples/artificial-algorithm/params.pcs`
See [Parameter Configuration Space File Format](#parameter-configuration-space-file-format)
for more details.

**Target Algorithm Running Time Cutoff:** The maximum amount of time GPS should
wait for a single target algorithm call to complete. Ideally, the default
configuration of your target algorithm should be able to solve at least 90%
of the problem instances within this running time cutoff. The performance of
all algorithm configurators is very sensitive to this parameter. If you are
unsure what value to use, you can use a very large one. GPS will adaptively
limit the running times of the target algorithm calls; however, it may 
initially waste considerable time if some of the first few target algorithm
calls it performs take extremely long to run, since it won't yet know what
a good value to use is. This corresponds to the `algo-cutoff-time` argument
and should be specified in seconds.
For example:
`--algo-cutoff-time 600`

**Configuration Budget:** You must specify a configuration budget using at 
least one of GPS's three configuration budget limits. These are:
 - `wallclock-limit` - Which specifies the total wall clock time allowed
for the GPS run.
 - `cputime-limit` - Which specifies the total amount of CPU time allowed
to be spent evaluating configurations (exlcuding overhead from GPS) for
the GPS run.
 - `runcount-limit` - Which specifies the total number of target algorithm
runs allowed for the GPS run.

You may use multiple configuration budget limits together. For example:
`--runcount-limit 400 --cputime-limit 14400`. Whichever one is exhausted
first will terminate the GPS run. 

**Redis Database Configuration:** You must also tell GPS how to connect to
your redis database server. Normally, you will want to specify both the 
redis host and port using the `redis_configuration.txt` file in the main
GPS directory, as these will typically not change between GPS runs. For
example, the file contents could be:

    redis-host = localhost
    redis-port = 9503

If you choose, you can also specify this information on the command line:
`--redis-host localhost --redis-port 9503`.

You will also need to speciy the redis database instance ID to be used
for the GPS run. You can perform multiple independent runs of GPS in
parallel, in which case each run must use a separate database ID. For
this reason, we recommend to always specify this value on the command line.
For example:
`--redis-dbid 0`.

## Example command line call for GPS

Combining all of the above examples, you're now ready to perform your first
run of GPS. From the base GPS directory, run:

    python2 run_gps_master.py --algo 'python2 examples/artificial-algorithm/algorithm.py' --instance_file examples/artificial-algorithm/instances.txt --pcs-file examples/artificial-algorithm/params.pcs --algo-cutoff-time 600 --runcount-limit 400 --cputime-limit 14400 --redis-dbid 0

This will setup the scenario files and output directory for the GPS run. It will
then stop and wait until there is at least 1 GPS worker ready to start. It should
print similar output to the following to the console:

    [INFO]:2020-07-02 11:39:05,801: Starting new GPS run with GPS ID 4RP76T
    [INFO]:2020-07-02 11:39:05,804: Waiting until all workers are ready...
    [INFO]:2020-07-02 11:39:06,806: There are 0 out of a minimum of 1 workers ready...

Next, in a second terminal, run 

    python2 run_gps_worker.py --redis-dbid 0

Note that you do not need to set any arguments other than the redis database
ID. The first process you started will have already created all of the files
necessary for the GPS run, including a new copy of the scenario file with all
GPS arguments fully instantiated. The worker will connect to the specified 
database to determine the location of these files, and then it will begin to
query the database for target algorithm runs to perform. At this time, the
original process will begin running the GPS master process. 

Normally, you would typically want to provide GPS with additional workers to
distribute the work of performing target algorithm runs. There is no limit
to the number of workers that GPS can use. You can continue to start as
many worker processes as you like. However, since this scenario uses an 
artificial algorithm that requires virtually no time to perform the target
algorithm runs, there is no benefit to doing so.

The entire process should take about 60-90 seconds to run, although it 
occassionaly requires up to 5 minutes. While running, GPS will print
information to the console on the current status of its run, including any
time it updates the value of a parameter in the incumbent configuration. 
When it is done, you should see output similar to the following printed to 
the console from the master process:

    [INFO]:2020-07-02 11:40:34,070: Reason for stopping: run budget exhausted
    [INFO]:2020-07-02 11:40:34,070: Used: 10289.0807883 CPU Seconds on target algorithm runs
    [INFO]:2020-07-02 11:40:34,071: Used: 67.2261619568 Wall Clock Seconds (total)
    [INFO]:2020-07-02 11:40:34,071: Used: 400 target algorithm runs.
    [INFO]:2020-07-02 11:40:34,071: Final Incumbent:  -heuristic 'a' -x0 '5' -x1 '0.999546896146'

However, since GPS is a randomized algorithm the exact output will vary. 
See `examples/artificial-algorithm/readme.txt` for more information about
this scenario.

## Experiment Directory

Rather than running everything from the GPS root directory, it is often more
convenient to specify the location from which the experiment should be run 
using the `experiment-dir` argument. When this is done, all other relative file 
and directory paths should be specified relative to this directory. Note that
GPS will then change to this directory prior to beginning to run, which means
that any relative paths used internally by your target algorithm or wrapper 
must be available from your experiment directory.

This is also a convenient way to help organize the several output files used
by GPS, as they will be stored in this experiment directory rather than the
directory from which you called GPS. To continue our running example, you
would start the master process with the following command line call:

    python2 run_gps_master.py --experiment-dir examples/artificial-algorithm/ --algo 'python2 algorithm.py' --instance_file instances.txt --pcs-file params.pcs --algo-cutoff-time 600 --runcount-limit 400 --cputime-limit 14400 --redis-dbid 0

The call for the worker process remains the same.

## Using a Scenario file

Rather than using a long command line call to start GPS, it is often helpful
to specify most of GPS's arguments in a scenario file, and then to simply
point GPS to the scenario file. For example, we include the file
`examples/artificial-algorithm/scenario.txt`, which contains the following
argument specifications:

    pcs-file = params.pcs
    algo = python algorithm.py
    algo_cutoff_time = 600
    instances = instances.txt
    # Whichever budget limit is reached first will terminate GPS
    runcount_limit = 400
    # Note that GPS only counts the times reported by your target algorithm in this limit
    # So even though we are giving it 6 hours (21600 seconds), it should actually terminate
    # in around 1-3 minutess, since our artificial algorithm actually spends far less time
    # than it returns
    cputime_limit = 21600
    verbose = 1

Any line that begins with `#` is treated as a comment and ignored. 

You can then start the GPS master process using:

    python2 run_gps_master.py --experiment-dir examples/artificial-algorithm --scenario-file scenario.txt --redis-dbid 0

The call for the worker process remains the same.

Any argument that can be specified on the command line can also be specified 
in a scenario file (except the scenario file itself). If the same argument is
defined multiple times, the order of precedence will be:
command line > scenario file > redis configuration file > GPS default values.

## Temporary File Directory - **Important**

GPS (like other algorithm configurators) creates a large number of temporary 
files that it uses to interact with your target algorithm wrapper (provided
you use the command line interface instead of the python interface). This can
sometimes cause temporary performance degredation for your entire filesystem
(which in turn, can of course impact the quality of the configurations found
by GPS). GPS will clean up these files when it is done with them. However, 
for file systems that automatically back-up files, GPS and other algorithm
configurators can still cause performance degradation. 

For this reason, it 
is **strongly** recommended to provide GPS with the location of a temporary
directory where it can create these files without effecting the performance
of the main filesystem. This directory should not be backed up, and should
ideally be fast to write to.

The temporary directory can be specified using the `temp-dir` argument. For
example: `--temp-dir /tmp`.

GPS does not share temporary files between
processes. So if worker processes are operating on separate nodes of a 
cluster, it is unproblematic for each worker to have access to separate
temporary file directories. 

## Solution Quality Optimization

In addition to minimizing the running times of a target algorithm, GPS can be
used to minimize the solution qualities found by the target algorithm, if 
applicable. For example, used this way GPS can optimize the validation error
of machine learning models through hyper-parameter optimization, or GPS can
be applied to optimize the performance of optimization algorithms, allowing
them to find better solutions within a fixed computational budget. Recent
evidence suggests that for scenarios where both metrics are available, 
optimizing for solution quality should provide configurators with a more
informative objective function, allowing them to find higher-quality
parameter configurations (see, *e.g.*, Hall *et al.*, 2020, "Analysis of the 
Performance of Algorithm Configurators for Search Heuristics with Global 
Mutation Operators").

An example artificial scenario showing how to use GPS for solution quality
optimization can be found in `./examples/artificial-classifier`,
which is designed to approximately resemble the behaviour of a machine learning
binary classification scenario. This scenario can be run with:

    python run_gps_master.py --experiment-dir examples/artificial-classifier --scenario-file scenario.txt --dbid 0

and

    python run_gps_worker.py --dbid 0

Apart from providing GPS with the solution quality to optimize, it 
is required to specify in the scenario file that the `run_obj` is `QUALITY`
instead of the default `RUNTIME`. When optimizing solution quality, GPS 
will minimize the mean solution quality.

Note that it does not make sense to use adaptive capping when performing
solution quality optimization. Therefore, adaptive capping will be disabled
and any values passed to the `bound_multiplier` parameter will be ignored. 


# Target Algorithm Wrapper

When performing automated algorithm configuration, it is typical to use a
target algorithm wrapper that implements a particular command line interface
between the algorithm configurator and the target algorithm. GPS supports both
the conventional command line interface, as well as a python interface.
The target algorithm wrapper is responsable for calling the target algorithm 
using the specified configuration on the specified instance, measuring the
runing time or solution quality (*e.g.*, validation loss), and enforcing the 
running time cutoff. 

GPS uses the same command line interface as SMAC and 
ParamILS, which means that you can directly use any scenarios set up for use
with the generic wrapper for algorithm configuration available from 
https://github.com/automl/GenericWrapper4AC. However, if your target algorithm
is implemented in python, then using the python interface will potentially 

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [1.1.1] - 2020-08-05

### Fixed
 - Previous version of GPS contained a bug causing any configuration that was censored a single time using
   the user-specified running time cutoff to be treated as if they had been censored due to an adaptive cap.
   For large running time cutoffs, this rarely affects the performance of GPS, in fact, in may provide a
   small speedup as GPS will reject bad configurations more quickly. However, if the running time cutoff is
   small enough to occasionally censor the runs of good configurations, then the output from GPS will likely
   be quite poor, since GPS will effectively be random rejecting configurations. This bug did not exist in the
   original version of GPS studied in the 2020 GECCO paper: "Golden Parameter Search: Exploiting Structure to
   Quickly Configure Parameters in Parallel".

## [1.1.0] - 2020-07-17

### Added
 - GPS now supports solution quality optimization! See `examples/artificial-classifier` for an example on how to use it. More information is available in `README.md`.
 - GPS now has a changelog.

## [1.0.1] - 2020-07-16

### Added
 - New instructions for setting up a redis database installation, provided by Marie Anastacio.
 - Several minor bugfixes and additions to the documentation.

## [1.0.0] - 2020-07-02

### Added
- First public-ready version of GPS!


GPS/pcsHelper.py
#Author: YP
#Created: 2019-03-07
#Last updated: 2019-04-25
#A collection of functions used by GPS that interacts with pcs objects.
#Also includes some handy functions for modifying PCS files into formats better
#for GPS

import copy as cp
from PCS import pcsParser


def reparameterizePCS(pcsFile):
    #Author: YP
    #Created: 2019-04-25

    pcs = pcsParser.PCS(pcsFile)

    for p in pcs.paramList:
        comment = pcs.getAttr(p,'comment')
        if(len(comment) > 0):
            comment = pcs.getAttr(comment,'text')
        if(pcs.getAttr(p,'type') == 'categorical' and 'treat numeric' in comment):
            #We are converting this number-based categorical parameter into a numerical parameter
            p = pcs.getObject(p)
            values = []
            integer = True
            for val in pcs.getAttr(p,'values'):
                val = pcs.getAttr(val,'text')
                try:
                    val = int(val)
                except:
                    try:
                        val = float(val)
                        integer = False
                    except:
                        val = val.strip()
                values.append(val)

            if(integer):
                p['type'] = 'integer'
            else:
                p['type'] = 'real'

            #Check if there are any values that need to be removed and treated
            #as a categorical parameter.
            removeValues = []
            if('strip values [' in comment):
                removeValues = comment.split('strip values [')[-1].split(']')[0].split(',')
                for i in range(0,len(removeValues)):
                    try:
                        removeValues[i] = int(removeValues[i])   
                    except:
                        try:
                            removeValues[i] = float(removeValues[i])
                        except:
                            removeValues[i] = removeValues[i].strip()

                for val in removeValues:
                    values.remove(val)

            #get the new range of the parameter.
            newValues = [min(values),max(values)]
            newDefault = pcs.getAttr(pcs.getAttr(p,'default'),'text')
            parentDefault = 'on'
            try:
                newDefault = int(newDefault)
            except:
                try:
                    newDefault = float(newDefault)
                except:
                    #The old default value is not one of the values left.
                    #We invent a new default value for this parameter by taking
                    #the middle of the parameter's range
                    newDefault = sum(newValues)/2
                    parentDefault = pcs.getAttr(pcs.getAttr(p,'default'),'text')
            if(newDefault in removeValues):
                #See comment above
                newDefault = sum(newValues)/2
                parentDefault = pcs.getAttr(pcs.getAttr(p,'default'),'text')

            p['default'] = newDefault
            p['values'] = newValues
            p['log'] = False
            
            if(len(removeValues) > 0):
                #We need to create a parent for the old parameter that contains 
                #as options "on" and the values of removedValues. It will then be
                #up to the wrapper to correctly use the value of the child
                #parameter if the parent is set to on, and to correctly replace
                #the value of the child parameter with the value specified in the 
                #parent, otherwise. 
                parentName = '__parent__' + pcs.getAttr(p,'name')
                parentValues = removeValues
                parentValues.append('on')
                parentValues = ', '.join([str(v) for v in parentValues])
                parentComment = '# Created automatically as a parent for ' + pcs.getAttr(p,'name')
                #Create a text-based representation of the new parent parameter
                line = parentName + ' categorical {' + parentValues + '} [' + parentDefault + '] ' + parentComment
                #Parse it
                obj = pcs.parseCategorical(line)
                #and add it to the document
                pcs.doc['content'].append(obj['id'])
                
                #Now create, parse and add the conditional statement.
                childName = pcs.getAttr(p,'name')                
                line = childName + ' | ' + parentName + ' == on # Created automatically to set the parent-child relationship'
                #parse it
                obj = pcs.parseConditional(line)
                #add it
                pcs.doc['content'].append(obj['id'])



 
        elif(pcs.getAttr(p,'type') == 'integer' and 'strip values [' in comment):
            #We only support removing either the smallest or the largest k elements
            removeValues = comment.split('strip values [')[-1].split(']')[0].split(',')
            for i in range(0,len(removeValues)):
               removeValues[i] = int(removeValues[i])   
            newValues = pcs.getAttr(p,'values')
            parentValues = cp.copy(removeValues)
            while(len(removeValues) > 0):
                if(min(removeValues) == newValues[0]):
                    newValues[0] += 1
                    removeValues.remove(min(removeValues))
                elif(max(removeValues) == newValues[1]):
                    newValues[1] -= 1
                    removeValues.remove(max(removeValues))
                else:
                    raise Exception("Unable to remove any more values because they are not at the edge of " + str(pcs.getAttr(p,'name')) + "'s range .")

            p = pcs.getObject(p)
            p['values'] = newValues

            if(pcs.getAttr(p,'default') in parentValues):
                parentDefault = pcs.getAttr(p,'default')
                newDefault = sum(newValues)/2
                p['default'] = newDefault
            else:
                parentDefault = 'on'
                

            #We need to create a parent for the old parameter that contains 
            #as options "on" and the values of removedValues. It will then be
            #up to the wrapper to correctly use the value of the child
            #parameter if the parent is set to on, and to correctly replace
            #the value of the child parameter with the value specified in the 
            #parent, otherwise. 
            parentName = '__parent__' + pcs.getAttr(p,'name')
            parentValues.append('on')
            parentValues = ', '.join([str(v) for v in parentValues])
            parentComment = '# Created automatically as a parent for ' + pcs.getAttr(p,'name')
            #Create a text-based representation of the new parent parameter
            line = parentName + ' categorical {' + parentValues + '} [' + str(parentDefault) + '] ' + parentComment
            #Parse it
            obj = pcs.parseCategorical(line)
            #and add it to the document
            pcs.doc['content'].append(obj['id'])
               
            #Now create, parse and add the conditional statement.
            childName = pcs.getAttr(p,'name')                
            line = childName + ' | ' + parentName + ' == on # Created automatically to set the parent-child relationship'
            #parse it
            obj = pcs.parseConditional(line)
            #add it
            pcs.doc['content'].append(obj['id'])


           

    #If any conditional statements use "in" we have to remove them
    #and create one copy of the child parameter and conditional statement for 
    #each parameter value in the "in" statement. The new conditional statements
    #will therefore only contain "==" as their operators. 
    for c in pcs.conditionList:
        clause = pcs.getAttr(c,'clauses') 
        operator = pcs.getAttr(clause,'operator')
        child = pcs.getAttr(c,'child')
        print("'" + operator + "'")
        if(operator == 'in'):
            #print("Working on this one")
            #Remove the old, single version of the child and conditional
            pcs.doc['content'].remove(pcs.getAttr(child,'id'))
            pcs.doc['content'].remove(pcs.getAttr(c,'id'))
            #Now we will create the new ones. 
            B = pcs.getAttr(pcs.getAttr(clause,'B'),'values')

            parentName = pcs.getAttr(pcs.getAttr(clause,'A'),'name')
            for val in B:
                valueText = pcs.getAttr(val,'text')
                oldChildName = pcs.getAttr(child,'name')
                childName = '__child_' + valueText + '__' + oldChildName
                #Create the new line for the new child
                line = pcs.printObject(child)[0].replace(oldChildName,childName) + " #Automatically created child for parameter value " + valueText + " of parent " + parentName
                #Parse it
                childType = pcs.getAttr(child,'type')
                if(childType == 'integer'):
                    obj = pcs.parseInteger(line)
                elif(childType == 'real'):
                    obj = pcs.parseReal(line)
                elif(childType == 'categorical'):
                    obj = pcs.parseCategorical(line)
                elif(childType == 'ordinal'):
                    obj = pcs.parseOrdinal(line)
                else:
                    raise Exception("Unsupported parameter type")
                #Append it to the document
                pcs.doc['content'].append(obj['id'])

                #Create the conditional statement
                line = childName + ' | ' + parentName + ' == ' + valueText
                #parse it
                obj = pcs.parseConditional(line)
                #append it
                pcs.doc['content'].append(obj['id'])
                   

                
    #Print out the new GPS-friendly version of the document
    with open(pcsFile[:-4] + '-gps.pcs','w') as f_out:
        f_out.write(pcs.printDocument())
       




def handleInactive(pcs,config,p):
    #Author: YP
    #Created: 2019-03-06
    #Last updated: 2019-04-25
    #Removes any parameters whose parents are set such that they should be inactive. 
    #The only exception to this is for parameter p. If it should be inactive, then we
    #change its parent(s) to make it active. 

    config = cp.deepcopy(config)

    #Check if p is active
    if(not pcs.isActive(p,config)):
        setActive(pcs,config,p)


    return removeInactive(pcs,config)


def removeInactive(pcs,config):
    #Author: YP
    #Created: 2019-04-26
    #Removes all inactive child parameters. 

    #Now reduce the configuration only to the remaining active parameters.
    #We need to keep on doing this until no changes occur, since there may
    #be grandchildren that are not turned off correctly in the first pass
    #that removes their parents. 
    oldConfig = config
    reducedConfig = {}
    changed = True
    while(changed):
        changed = False
        for p in oldConfig.keys():
            if(pcs.isActive(p,oldConfig)):
                reducedConfig[p] = oldConfig[p]
            else:
                changed = True
        oldConfig = reducedConfig
        reducedConfig = {}

    return oldConfig


def setActive(pcs,config,p):
    #Author: YP
    #Created: 2019-04-25
    #Last updated: 2019-04-25
    #Recursively sets any parents to be active as necessary. 
    p = pcs.lookupParamID(p)
    conds = pcs.getParentConditions(p)
    for cond in conds:
        paramUpdates, valueUpdates = getClauseAssignments(pcs,pcs.getAttr(cond,'clauses'),pcs.convertConfigToIdsAndText(config))
        checkParents = set([])
        for i in range(0,len(paramUpdates)):
            config[paramUpdates[i]] = valueUpdates[i]
            checkParents.add(paramUpdates[i])

    #Check if we also need to turn on some grandparents
    for parent in checkParents:
        if(not pcs.isActive(parent,config)):
            setActive(pcs,config,parent) 
 
    
            
def getClauseAssignments(pcs,obj,config):
    #Author: YP
    #Created: 2019-03-07
    #Last updated: 2019-03-07
    #Evaluates the condition using the configuration specified in config.
    #Config must be a dict with parameters as keys (ids or objects), and 
    #the values must be the parameter values (either as ids or objects)
    #For any parts of the clause that would evaluate to False, the config
    #is minimally updated until the clause returns true. 
    #Note that this functions assumes that all condition operators are 
    #either ==,>=,<= and &&. None others are permitted.
    #Also note, that the convention must always be "parameter operator 
    #value". The only exception to this is for &&, in which case it is
    #"clause && clause".
    #NOTE: Update: I just learned that SMAC doesn't actually support the
    #"<=" and ">=" operators in conditionals. However, since we might want
    #to later, I'm going to leave the code here for them. the pcsParser 
    #will need to be extended to handle them, though. 

    obj = pcs.getObject(obj)
    if(pcs.isParameter(obj)):
        #The object is a parameter, so we return the value
        #for the parameter and the parameter
        if(obj['id'] not in config.keys()):
            raise Exception("There is not enough information about the parameter configuration to determine if the parameter should be active.")
        return config[obj['id']], obj['name']
    elif(pcs.getAttr(obj,'type') == 'value'):
        return pcs.getAttr(obj,'text')
    elif(obj['type'] == 'clause'):
        #The object is a clause, so we need to evaluate it (possibly 
        #using recursion)
        operator = obj['operator']
        if(operator == '&&'):
             aParamUpdates, aValueUpdates = getClauseAssignments(pcs,obj['A'],config)
             bParamUpdates, bValueUpdates = getClauseAssignments(pcs,obj['B'],config)
             aParamUpdates.extend(bParamUpdates)
             aValueUpdates.extend(bValueUpdates)
             return aParamUpdates, aValueUpdates
        elif(operator in ['<=','>=','<','>']):
            #We don't support ordinals here, so this won't handle them
            #correctly.
            A, Aname = getClauseAssignments(pcs,obj['A'],config)
            A = float(A)
            B = float(getClauseAssignments(pcs,obj['B'],config))
            if(operator == '<='):
                if(not A <= B):
                    return [Aname], [B]
                else:
                    return [], []
            elif(operator == '>='):
                if(not A >= B):
                    return [Aname], [B]
                else:
                    return [], []
            else:
                raise Exception("Invalid operator")
        elif(operator in ['==']):
            A, Aname = getClauseAssignments(pcs,obj['A'],config)           
            B = getClauseAssignments(pcs,obj['B'],config)
            #A and B are now values as strings
            if(not A == B):
                return [Aname], [B]
            else:
                return [], []
           
    raise Exception("Operator not supported.")




